### ISSUE TITLE
Amplify codegen models on Datastore cannot generate Type with no @model directive

### ISSUE THREAD

**Issue by andikawhy** [aws-amplify/amplify-cli#2988]
**Describe the bug**
I have a schema with no @model directive, it's only use as typings on my typescript project. When I tried amplify datastore `amplify codegen models`, cli says 

```
Unknown type Content for field value. Did you forget to add the @model directive
Error: Unknown type Value for field value. Did you forget to add the @model directive
```

**Amplify CLI Version**
4.6.0

**To Reproduce**
Simply create a schema.graphql like this
```
type Post @model {
  id: ID!
  title: String!
  content: Content!
  price: Int
  rating: Float
}

type Content {
  id: String
  name: String
}
```
then run `amplify codegen models` to generate the model.

**Expected behavior**
No error.

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Node Version: 10.16.0

**Comment by idanlo** [comment]
+1

**Comment by yonatanganot** [comment]
+1

**Comment by sacrampton** [comment]
I have the same issue - my use case is for geolocations (lat/lon) that are passed to ElasticSearch as a geopoint for geospatial queries. Without this being fixed we can't move forward.

**Comment by mochfamir** [comment]
+1

**Comment by Soek22** [comment]
I think this error is arising because 'codegen models' doesn't generate code for types without `@model` directive, so when it tries to create a field with that type, it doesn't find the 'content' type.
I have reported this behavior as a bug: https://github.com/aws-amplify/amplify-cli/issues/3026

**Comment by yuth** [comment]
Thank you for bringing this issue up. We are actively evaluating how we can add support non model types in DataStore.

**Comment by sacrampton** [comment]
I am thinking of having different versions of the schema.graphql to get around this issue. Can you advise whether this is valid?

**Version 1 - Used for amplify push (generating DynamoDB)**
```
type Post @model {
  id: ID!
  title: String!
  content: Content!
  price: Int
  rating: Float
}

type Content {
  id: String
  name: String
}
```
**Version 2 - Used for amplify codegen models**
```
type Post @model {
  id: ID!
  title: String!
  content: Content!
  price: Int
  rating: Float
}

type Content @model {
  id: String
  name: String
}
```

If I do that then will there be any issue when it comes to syncing the model with the cloud?

**Comment by Soek22** [comment]
@sacrampton I think both will overlap somewhere. The approach I am following is: 

1. Generate the code with @model directive: `$amplify codegen models`, and import it to the project (manually so later it doesn't get overridden).
2. Remove the @model directive, and `$amplify api push`.

I still couldn't test it as I am migrating other code from the previous SDK. Also I am using custom resolvers that can make it a little bit trickier until I manage to confirm if it works.

But you can give it a try.

**Comment by sacrampton** [comment]
Thanks @Soek22  - just to make sure I'm not misunderstanding you - with your item #1 above, when you say import the models into the project manually - can you explain what you are doing there. Does this mean you generate the models in a separate project (ie. \user\models) then manually copy the models into your appsync project directory (ie. \user\appsync)

I think that is what you are saying, but I just want to be sure.

**Comment by Soek22** [comment]
@sacrampton I am using iOS target at this moment, but I suppose it works the same way with other target platforms. When I say import I mean import the code to Xcode (iOS IDE).

1. First I set my graph with @model directives and generate front-end code. It can be generated with the amplify CLI (`$amplify codegen models`) or with specific method for each target platform:
Once generated I get front-end code for my target platform ([iOS](https://aws-amplify.github.io/docs/ios/api#generate-your-model-files)) on `project-root-directory/amplify/generated/models`(`app/src/main/java/com/amplifyframework.datastore.generated.model` on [Android](https://aws-amplify.github.io/docs/android/start#step-2-generate-your-model-files), and `./models` on [Javascript](https://aws-amplify.github.io/docs/js/datastore#model-generation)). This is done locally (on your machine without any was backend involved) through the codegen.

2. Then I copy manually those files to a safe path or rename them (so in next codegen executions doesn't get overridden), and import them to my project.
-  1. iOS: import them to Xcode and add them as target for build time.
- 2. Android: IDK
- 3. [Javascript](https://aws-amplify.github.io/docs/js/datastore#setup): `import { Post, Content } from "./safe-models";`
*When I say import, I mean import them to the app code, with your IDE/whatever. As this is the front-end code to be used when compiling your app.

3. Remove the @model directive, and $amplify api push.

In case this works is just a workaround, as requires manual intervention every time you update your schema.

**Comment by rpostulart** [comment]
+1

Here is my schema:

type Event @model @searchable {
  id: ID!
  date: String!
  maxSubscribers: Int
  totalSubscribers: Int
  polygon: Location
  Timetable: [Timetable]
  lastDateToSubscribe: String
  eventOwner: Subscriber! @connection
  name: String!
  description: String
  coordinates: [Coordinate]
  planned: Boolean
  subscribers: [EventSubscribers] @connection(name: "Event")
}

type EventSubscribers
  @model(queries: null)
  @key(fields: ["eventID", "createdAt"]) {
  eventID: String!
  createdAt: String!
  subscriber: Subscriber! @connection(name: "EventSubscriber")
  event: Event! @connection(name: "Event")
  totalPersons: Int!
}

type Subscriber @model {
  id: ID!
  name: String
  email: String
  phone: String
  device: String
  expoToken: String
  addressLocation: Location
  address: String
  complete: Boolean
  events: [EventSubscribers] @connection(name: "EventSubscriber")
  planning: [Planning] @connection(name: "SubscriberPlanning")
}

type Planning @model {
  id: ID!
  event: Event
  subscriber: Subscriber @connection(name: "SubscriberPlanning")
  location: Location
  description: String
  type: String
  time: String
  notificationSend: Boolean
}

type Location {
  type: String
  coordinates: [[[Float]]]
}
type Coordinate {
  longitude: Float
  latitude: Float
}

type Timetable {
  description: String
  startTime: String
  endTime: String
}

input LocationInput {
  type: String
  coordinates: [[[Float]]]
}

type EventConnection {
  items: [Event]
  total: Int
  nextToken: String
}

input PointInput {
  type: String
  coordinates: [Float]
}

type Query {
  nearbyEvent(location: PointInput!): EventConnection
}

type Mutation {
  sendNotification(input: NotificationInput): NotificationResult
    @function(name: "sendNotifications-${env}")
  createPayment(input: PaymentInput): PaymentResult
    @function(name: "makePayment-${env}")
}

input PaymentInput {
  amount: Float
  total: Int
}

type PaymentResult {
  statusCode: Int
  body: String
}

type NotificationResult {
  result: String
}
input NotificationInput {
  action: String
  data: String
}

**Comment by sacrampton** [comment]
Hi @rpostulart - separate issue, I've found that Elasticsearch hard codes geopoint types with lat/lon coordinate names if you want to be able to do geo-queries

type Coordinate {
lat: Float
lon: Float
}

**Comment by idanlo** [comment]
Any update on this?

**Comment by HaseldenOrganizationOwner** [comment]
I've been able to work around this issue using JSON.parse and JSON.stringify in application code and just storing JSON strings in the database, at least until support is provided for complex object types that don't use @model

**Comment by mdebo** [comment]
HI, I meet exactly the same issue, any update about this? 
This feature is really important to allow the migration of existing API to datastore

**Comment by danrivett** [comment]
@mdebo - My workaround is still the one outlined by @lazy-var in https://github.com/aws-amplify/amplify-cli/issues/2988#issuecomment-567711467 to codegen with the @model annotations on all nested-types but then remove them before doing an `amplify push`. 

Not ideal in the long run but at least for the the short term it's an option.

**Comment by mdebo** [comment]
@danrivett thank you, i'm going to give it a try.

**Comment by jackatkoeln** [comment]
+1  because the workaround is very annoying :-(

**Comment by jackatkoeln** [comment]
> @mdebo - My workaround is still the one outlined by @lazy-var in [#2988 (comment)](https://github.com/aws-amplify/amplify-cli/issues/2988#issuecomment-567711467) to codegen with the @model annotations on all nested-types but then remove them before doing an `amplify push`.
> 
> Not ideal in the long run but at least for the the short term it's an option.

This will still not work with datastore. 
https://github.com/aws-amplify/amplify-js/issues/4579

**Comment by AndrewPrifer** [comment]
@danrivett I am a little confused since this shouldn't be working because DataStore doesn't support nested objects at the moment. Did you verify that this workaround works? If yes, are there any caveats?

**Comment by sacrampton** [comment]
My concern with this stuff is geolocations - the documentation tells you how to create a custom geolocation search resolver that creates ElasticSearch GeoPoints. I'm not sure this works with DataStore and I'm not sure there is any work around to this.

https://aws-amplify.github.io/docs/cli-toolchain/graphql#add-a-custom-geolocation-search-resolver-that-targets-an-elasticsearch-domain-created-by-searchable

The workaround is saying have this when doing codegen for models
```
type Location @model {
  lat: Float
  lon: Float
}
```

Then have this when doing amplify push...
```
type Location {
  lat: Float
  lon: Float
}
```

Is this even valid?

**Comment by dyson-yamashita** [comment]
I have same problem about nested objects . 
So, I changed from Amplify iOS  to Amplify iOS SDK(AppSync).
https://aws-amplify.github.io/docs/sdk/ios/start

**Comment by danrivett** [comment]
> @danrivett I am a little confused since this shouldn't be working because DataStore doesn't support nested objects at the moment. Did you verify that this workaround works? If yes, are there any caveats?

Let me get back to you as I was doing some POC work around this about a month ago and although what I suggested worked I didn't complete the POC since I was waiting on the React Native support to go live since it was an RN POC.

Therefore it could have been premature of me to confirm that initial suggestion since I haven't tested syncing data with the backend, I stopped at confirming `amplify codegen` and `amplify push` worked successfully.

**Comment by AndrewPrifer** [comment]
@danrivett I see. It makes sense that the two operations work independently, however since they result in two completely different database layouts and GraphQL schemas, I'd be surprised if syncing worked. If you somehow manage to get it to work with any kind of data, I'd be very interested to know.

**Comment by jackatkoeln** [comment]
> My concern with this stuff is geolocations - the documentation tells you how to create a custom geolocation search resolver that creates ElasticSearch GeoPoints. I'm not sure this works with DataStore and I'm not sure there is any work around to this.
> 
> https://aws-amplify.github.io/docs/cli-toolchain/graphql#add-a-custom-geolocation-search-resolver-that-targets-an-elasticsearch-domain-created-by-searchable
> 
> The workaround is saying have this when doing codegen for models
> 
> ```
> type Location @model {
>   lat: Float
>   lon: Float
> }
> ```
> 
> Then have this when doing amplify push...
> 
> ```
> type Location {
>   lat: Float
>   lon: Float
> }
> ```
> 
> Is this even valid?

@sacrampton This is not a valid option. Datastore doesn't support nested types yet. If you use `codegen` with `@model` but then remove it for `amplify push` you will end up with different structures. I have tested it, you can still use Datastore.save locally and it works but it will not synchronize via apsync.  If it helps I can post the error messages in this thread.

**Comment by AndrewPrifer** [comment]
@jackatkoeln thanks for confirming it.

**Comment by danrivett** [comment]
@yuth @undefobj I would _really_ appreciate an update on this issue as something we're impacted by. The last update from AWS as far as I can see was back on December 18th and there's a number of us in this ticket that this is causing a problem for - primarily to handle structured geo data.

Be great to understand if this is something that is likely to be addressed in the next couple of months so we can understand how much effort we should put in to work around this issue.

**Comment by yuth** [comment]
We’re actively working on this, while we cannot guarantee a date we hope to have something in the coming weeks or two

**Comment by danrivett** [comment]
That's perfect, thanks @yuth. Dates are never as we expect, but knowing it is something that's made the cut for being worked on is a great update in and of itself. Thanks.

**Comment by lenarmazitov** [comment]
@yuth what about sorting functional, will it be presented in upcoming release? (sorry for off topic)

**Comment by slatemates** [comment]
We are facing a similar issue when doing amplify codegen models . Codegen complains about 
" Did you forget to add the @model directive " on the nested types. We are eagerly waiting for updates on this one . Thanks

**Comment by undefobj** [comment]
Wanted to provide an update to all customers on this thread. As @yuth mentioned we've been working on this and have a couple of releases in flight. You can see PRs here:

https://github.com/aws-amplify/amplify-js/pull/5128
https://github.com/aws-amplify/amplify-cli/pull/3718

We'll update soon when releases are ready for testing.

**Comment by rpostulart** [comment]
Cool! I am looking forward! Thanks already

**Comment by yuth** [comment]
Amplify CLI has added support for types without `@model` directive version 4.17.2. If you encounter any issues with model generation, please open an new issue.

**Comment by undefobj** [comment]
Also support for this is released in the latest version of Amplify JavaScript. Please have a test.

**Comment by danrivett** [comment]
This is huge! I will be testing this out shortly and let you know if I find anything. Great work guys.

**Comment by jackatkoeln** [comment]
I can confirm the changes, all working and syncing in all devices :-)

Simple test for the geolocation case requested by @sacrampton + some other directives for testing purpose. 

```graphql
type Restaurant
  @model
  @auth(rules: [{allow: private}]) {
  id: ID!
  name: String!
  guests: [Guest] @connection(name: "RestaurantGuests")
  rating: Int!
  location: Location
}

type Location {
  lat: Float
  lon: Float
}

type Guest
  @model
  @auth(rules: [{allow: private}]) {
  id: ID!
  name: String!
  restaurant: Restaurant @connection(name: "RestaurantGuests")
}
```
Boom! Thanks all for this much waited request. 🎉🚀

**Comment by danrivett** [comment]
Will it work with TypeScript ok btw? Just wondering if the correct type definitions are created for this? I'm not quite ready to play with this as have some inflight work, so thought I'd just check ahead of time.

**Comment by Computr0n** [comment]
I updated my amplify CLI and then after I removed the @model tags from my nested types and reran the modelgen and push I could not longer find references to the classes in my kotlin code. 
Is java codegen working for these nested types or am I missing something?

**Comment by danrivett** [comment]
@undefobj could you confirm if support was added to React Native environments? I assumed it was because I didn't see it mentioned it was only supported in certain environments, and this ticket has been closed off, but your comment and my afternoon of testing today seems to prove otherwise?

I'm currently seeing the following error in my log and no data synced. I blew away my non-prod environment and recreated it and still no luck:

```
[WARN] 06:11.431 DataStore, Object {
  "_err": [Error: Field description should be of type string, undefined received. undefined],
}
- node_modules/expo/build/environment/muteWarnings.fx.js:18:23 in warn
* [native code]:null in warn
- node_modules/@aws-amplify/core/lib/Logger/ConsoleLogger.js:85:8 in prototype._log
* http://127.0.0.1:19001/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&minify=false&hot=false:185768:22 in <unknown>
- node_modules/@aws-amplify/datastore/lib/sync/processors/mutation.js:264:52 in __generator$argument_1
* http://127.0.0.1:19001/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&minify=false&hot=false:227825:25 in step
- node_modules/@aws-amplify/datastore/lib/sync/processors/mutation.js:17:39 in rejected
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in callFunctionReturnFlushedQueue
```

**Comment by danrivett** [comment]
Ok good news, I'm not sure what happened as I deleted the API and recreated it before, but something was out of sync, and after further schema modifications and more `amplify codegen models` and `amplify push` commands later, and it's all working and syncing in RN 🎉 

Panic over, and overarching excitement back again!

**Comment by egoossens** [comment]
Seems to work fine. Thanks a lot for the release!

One question though, I was trying to have a schema like
```
type Post @model {
  id: ID!
  something: Something!
}

type Something {
  someField: String!
}
```
The something field on the Post type didn't have a ```@connection``` on purpose as I wanted to store it as a sub-document . The codegen warned me that I forgot the ```@connection``` directive and didn't generate the something property in the Post TypeScript class. Changing the field's name to something else removed the warning and the TypeScript class was generated as expected. Just curious if this is by design?

**Comment by yuth** [comment]
> Seems to work fine. Thanks a lot for the release!
> 
> 
> 
> One question though, I was trying to have a schema like
> 
> ```
> 
> type Post @model {
> 
>   id: ID!
> 
>   something: Something!
> 
> }
> 
> 
> 
> type Something {
> 
>   someField: String!
> 
> }
> 
> ```
> 
> The something field on the Post type didn't have a ```@connection``` on purpose as I wanted to store it as a sub-document . The codegen warned me that I forgot the ```@connection``` directive and didn't generate the something property in the Post TypeScript class. Changing the field's name to something else removed the warning and the TypeScript class was generated as expected. Just curious if this is by design?

This is not by design. Could you please file an separate issue and I will investigate what’s going on

**Comment by egoossens** [comment]
Hmm it seems I can't reproduce it now. Maybe I missed something last time. I'll open an issue if it happens again.

**Comment by ianmartorell** [comment]
Is this supposed to be working with `Interface`s?  I'm getting the same error:
```
Unknown type Block for field blocks. Did you forget to add the @model directive
Error: Unknown type Block for field blocks. Did you forget to add the @model directive
```
Schema:
```graphql
type Page @model {
  id: ID!
  ...
  blocks: [Block]
}

interface Block {
  id: ID!
}

type TextBlock implements Block {
  id: ID!
  title: String
  body: String
}
```

**Comment by lenarmazitov** [comment]
> Is this supposed to be working with `Interface`s? I'm getting the same error:
> 
> ```
> Unknown type Block for field blocks. Did you forget to add the @model directive
> Error: Unknown type Block for field blocks. Did you forget to add the @model directive
> ```
> 
> Schema:
> 
> ```graphql
> type Page @model {
>   id: ID!
>   ...
>   blocks: [Block]
> }
> 
> interface Block {
>   id: ID!
> }
> 
> type TextBlock implements Block {
>   id: ID!
>   title: String
>   body: String
> }
> ```

You trying to use interface Block as type in Page type

```
  blocks: [Block]
```

**Comment by ianmartorell** [comment]
> > Is this supposed to be working with `Interface`s? I'm getting the same error:
> > ```
> > Unknown type Block for field blocks. Did you forget to add the @model directive
> > Error: Unknown type Block for field blocks. Did you forget to add the @model directive
> > ```
> > 
> > 
> > Schema:
> > ```graphql
> > type Page @model {
> >   id: ID!
> >   ...
> >   blocks: [Block]
> > }
> > 
> > interface Block {
> >   id: ID!
> > }
> > 
> > type TextBlock implements Block {
> >   id: ID!
> >   title: String
> >   body: String
> > }
> > ```
> 
> You trying to use interface Block as type in Page type
> 
> ```
>   blocks: [Block]
> ```

I know, I'm doing that on purpose. Is it not supported? Schema generation and working with the API seems to be working fine.

**Comment by hellaeon** [comment]
@ianmartorell hi mate, I have just run into something similar - may I ask what you did to get past the 'interface' issue ?

**Comment by ianmartorell** [comment]
> @ianmartorell hi mate, I have just run into something similar - may I ask what you did to get past the 'interface' issue ?

I reformulated my schema to avoid using interfaces. 😅 I used a discriminator field instead in the model that would've been the interface.

**Comment by nielswart** [comment]
I am also unable to get model generation working with interface or Union types

I have the following schema

```
interface Action {
 ...
 }
type MoveRecord implements Action @model {
... 
}

type ActionRecord implements Action @model {
...
}

union Activity = ActionRecord | MoveRecord

type Group {

activity: [Activity]
}
```

It's very unfortunate that this does not work...

**Comment by rlee1990** [comment]
I can't get this to work when using this with flutter it does not create the object

**Comment by NimaSoroush** [comment]
Any update on this for Flutter apps?

**Comment by Alexminator99** [comment]
> Any update on this for Flutter apps?

I got the same problem now in Flutter. Awesome!!!

type S3Object {
  bucket: String!
  region: String!
  key: String!
}

type DocumentsPicture @model (...) {
  id: ID!
  name: String
  owner: String
  timestamp: AWSDate
  file: S3Object!
}

It only generates the licence header and import 2 .dart archives. Thats it for S3Object. Is there is a workaround?

Im in latest version -> 4.48.0. Want this issue to be reopened.
